### PR TITLE
[AOT] add option to control arguments to llc and add fallback for x86

### PR DESCRIPTION
**Pull request by lewurm**
Fixes https://github.com/mono/mono/issues/12182

### CODE REVIEW

**Comment by filipnavara** [comment]
Wouldn't this make the AOT binaries unusable if I transfer them to different machine with less capable CPU?

**Comment by vargaz** [comment]
Yes, that is the reason we don't use it. Perhaps we can add an aot command line option, or just the existing mtriple option.

**Comment by lewurm** [comment]
it isn't worse than before in that regard

**Comment by filipnavara** [comment]
My CPU supports AVX2. With the added flag it would pick it up and stop working even on SSE4.1 CPUs. That looks worse to me.

**Comment by lewurm** [comment]
That specific example, yes. In general it was broken in that way before, but I get your point 🙂 


I updated the PR, what do you think now?

@directhex it would mean you would need to update debian packaging so that it will pass `--aot=llvmllcopts=-mcpu=native` to the AOT compiler.

**Comment by filipnavara** [comment]
Realistically SSE4.1 was available on Intel CPUs for 11 years (Core 2 / Penryn) and on AMD CPUs for 8 years (AMD FX / Opteron / Bulldozer) so the base target was already quite low. I personally don't mind changing it as long as it does not break binaries I distribute to customers.

On some targets like macOS the base level could certainly be higher because the supported macOS version has already pretty high bar on CPU requirements. Currently Mono runs on macOS Mavericks or newer and Apple officially supports macOS Sierra and newer. The only machine without SSE4.1 where macOS Mavericks might have been supported is some low configuration of iMac from 2007 (Merom architecture) and I am not even sure if that one was officially supported or not.

File: mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -7873,6 +7884,8 @@ mono_aot_parse_options (const char *aot_options, MonoAotOptions *opts)
 			opts->llvm_opts = g_strdup (arg + strlen ("llvmopts="));
 		} else if (str_begins_with (arg, "llvmllc=")){
 			opts->llvm_llc = g_strdup (arg + strlen ("llvmllc="));
+		} else if (str_begins_with (arg, "llvmllcopts=")) {
+			opts->llvm_llc_opts = g_strdup (arg + strlen ("llvmllcopts="));
```
Comment (vargaz) [inline mono/mini/aot-compiler.c]: I would call this llcmopts. Alternative, we can add a mcpu/march argument similar to the mtuple argument.

**Comment by directhex** [comment]
@lewurm that's easily changed

Bear in mind 1) we had a bunch of angry bug reports recently because SSE4.1 as a hard dependency for LLVM AOT broke lots of people 2) the Debian/Ubuntu packages generate AOT caches at install time (not shipped in packages) so we can be as aggressive in optimizing as we like, as long as it decides what to do at install time (i.e. yay for `-mnative`)

**Comment by lewurm** [comment]
@directhex how far do you want this to be backported? 2018-10?

**Comment by lewurm** [comment]
@monojenkins backport 2019-02

**Comment by marek-safar** [comment]
@monojenkins backport 2018-10

**Comment by lewurm** [comment]
@marek-safar: @directhex communicated (via Slack) that we should only backport it to `2019-02`. It requires changes in Linux packaging.

**Comment by directhex** [comment]
So, what is the desired outcome of this work?

Right now, when you install a package with a `installmonoaot` manifest, this script gets executed:

https://github.com/mono/linux-packaging-mono/blob/b7e7a350b3d0dccf7287c72598be9aa8018d8031/debian/postinst-monoaot#L3-L25

So _IF_ `/proc/cpuinfo` contains `sse4_1`, and `/usr/lib/libmono-llvm.so.0` exists, it calls:
```
/usr/bin/mono --aot=try-llvm,temp-path=/tmp,outfile=/usr/lib/mono/aot-cache/${MONOARCH}/foo.dll.so` -O=all foo.dll
```

else it calls
```
/usr/bin/mono --aot=temp-path=/tmp,outfile=/usr/lib/mono/aot-cache/${MONOARCH}/foo.dll.so` -O=all foo.dll
```

What is the new desired behaviour? I guess the SSE4.1 check can be removed entirely? Are we comfortable that this will do the right thing on ARM etc?

**Comment by lewurm** [comment]
The script should pass `--aot=llvmllcopts=-mcpu=native` on i386/amd64.

For arm/aarch64: I'm not sure. It looks like LLVM would support `-mcpu=native` for those targets as well, but I have no idea how well it works in practice.
https://github.com/llvm/llvm-project/blob/4218d45f6581fd508f493522bf514981d5c9ca2b/clang/lib/Driver/ToolChains/Arch/ARM.cpp#L570-L583
https://github.com/llvm/llvm-project/blob/4218d45f6581fd508f493522bf514981d5c9ca2b/clang/lib/Driver/ToolChains/Arch/AArch64.cpp#L39-L43

**Comment by EgorBo** [comment]
We now have two ways of passing args to `llc`:
`--aot=llvm,llcopts=""`
and
`--aot=llvm,llvmllc=""`
maybe we should drop one? or why `llvmllc` was not changed to just override `-mcpu`?

**Comment by lewurm** [comment]
@EgorBo I falsely assumed that `--aot=llvmllc=[...]` would specify the path of `llc` 🤦‍♂️  could you please review https://github.com/mono/mono/pull/13877 ?

**Pull request merged**